src/models/call.rs

```diff
@@ -0,0 +1,266 @@
+use std::slice::SliceIndex;
+
+use bytes::Buf;
+use reth_primitives::{Bytes, Transaction, TxEip1559, TxEip2930, TxLegacy};
+use reth_rlp::{Decodable, Header};
+use starknet::accounts::Call as StarknetCall;
+use starknet_crypto::FieldElement;
+
+use crate::models::errors::ConversionError;
+use crate::starknet_client::helpers::DataDecodingError;
+
+#[derive(Clone)]
+pub struct Call(StarknetCall);
+
+impl From<Call> for StarknetCall {
+    fn from(call: Call) -> Self {
+        call.0
+    }
+}
+
+impl From<StarknetCall> for Call {
+    fn from(call: StarknetCall) -> Self {
+        Self(call)
+    }
+}
+
+impl From<Call> for Vec<FieldElement> {
+    fn from(call: Call) -> Self {
+        let mut c = vec![
+            FieldElement::ONE,
+            call.0.to,
+            call.0.selector,
+            FieldElement::ZERO,
+            FieldElement::from(call.0.calldata.len()),
+            FieldElement::from(call.0.calldata.len()),
+        ];
+        c.extend(call.0.calldata);
+        c
+    }
+}
+
+pub struct Calls(Vec<StarknetCall>);
+
+impl From<Calls> for Vec<StarknetCall> {
+    fn from(calls: Calls) -> Self {
+        calls.0
+    }
+}
+
+impl From<Vec<StarknetCall>> for Calls {
+    fn from(calls: Vec<StarknetCall>) -> Self {
+        Self(calls)
+    }
+}
+
+/// Converts a raw starknet transaction calldata to a vector of starknet calls.
+impl TryFrom<Vec<FieldElement>> for Calls {
+    type Error = ConversionError;
+
+    fn try_from(value: Vec<FieldElement>) -> Result<Self, Self::Error> {
+        // in account calls, the calldata is first each call as {contract address, selector, data offset,
+        // data length} and then all the calldata of each call, so each call takes 4 felts, and
+        // eventually the calldata of the first call is at offset =  1 (for call_len) + 4 * call_len + 1
+        // (for calldata_len)
+        let calls_len = u32::try_from(value[0])
+            .map_err(|e| ConversionError::ValueOutOfRange(format!("{}: call array length > u32::MAX", e)))?
+            as usize;
+
+        let mut offset = calls_len * 4 + 2;
+
+        let mut calls = vec![];
+        for i in 0..calls_len {
+            let calldata_len =
+                u32::try_from(value[i * 4 + 4]).map_err(|e| ConversionError::ValueOutOfRange(e.to_string()))? as usize;
+            let call = StarknetCall {
+                to: value[i * 4 + 2],
+                selector: value[i * 4 + 3],
+                calldata: value[offset..offset + calldata_len].to_vec(),
+            };
+            offset += calldata_len;
+            calls.push(call);
+        }
+        Ok(Self(calls))
+    }
+}
+
+impl TryFrom<Call> for Transaction {
+    type Error = DataDecodingError;
+
+    fn try_from(value: Call) -> std::result::Result<Self, Self::Error> {
+        let call = value.0.calldata.iter().filter_map(|x| u8::try_from(*x).ok()).collect::<Vec<u8>>();
```

Comment:
can we use `into_iter` and avoid the copy on x?

Reply:
can try